### ISSUE TITLE
Redesign AJ "fetch run" mechanism.

### ISSUE THREAD

**Issue by clevengr** [pc2ccs/pc2v9#496]
**Is your feature request related to a problem?** 

Yes. The current implementation, where AutoJudges "request" a specific run, has been determined to have numerous flaws (including that multiple AJs configured to judge the same problem(s) frequently end up requesting the same run, causing one of them to get the run and the others to have wasted time and now needing to go back and request a different run -- which might end up also being requested by other AJs, etc.  

**Feature Description**: 

Redesign the "AJ run acquisition" mechanism so that instead of an AJ requesting a specific run, AJs "register" as "AVAILABLE_TO_JUDGE", and on the server side when a run is received the server examines the list of "Available AJs" (restricting itself to only those that are configured to judge that particular problem), and then operate as if a "Run Checkout Request" had been received -- i.e., sending the received run to the selected AJ.

**Have you considered other ways to accomplish the same thing?** 
Other than staying with the current, flawed, implementation, nothing comes to mind.

**Do you have any specific suggestions for how your feature would be ***implemented*** in PC^2?** If so, please describe.  

- Add a new PacketType AJ_AVAILABLE.
- Have each AJ send such a packet when it starts up (instead of searching for a run to request).
- Have each AJ send such a packet when it _finishes_ judging a run (i.e. after sending a JUDGEMENT packet).
- Have the server maintain a list "AvailableAJs" of AJs which have sent an AJ_AVAILABLE packet.
- Have the server implement a loop such as
```
while (there are runs to be judged) {
  Search the availableAJs list for a judge configured for the problem;
  if (found) {
      Remove the AJ from the availableAJs list;
      Checkout the run and send it to the selected AJ;
  }
}
```
(The above is high-level pseudo-code for the concept; there are lots of details such as synchronization, exactly where the above loop belongs, what to do if there's no AJ available, etc.)

**Comment by lane55** [comment]
I disagree with using a loop.   The processing for a new run or AJ_AVAILABLE should
have individual logic/code, rather than having to debug code within a loop.   These
are isolated events and should not be clumped into a while loop, unless there is a
condition that requires a loop.

**Comment by lane55** [comment]
The solution should include a way for an Admin to view the AJ Lists, either via a live monitoring (unlikely, large LOE)
or a snapshot where the Admin requests the list contents and the server sends a packet with the list contents.

**Comment by lane55** [comment]
On server startup, there would be a loop to go through runs.

**Comment by lane55** [comment]
> Have the server maintain a list "AvailableAJs" of AJs which have sent an AJ_AVAILABLE packet.
this includes changing the list when the following events happen:
1. The AJ is logged off  
2. The AJ has had its autojudging turned off
3. As stated above, if a judge is configured as a auto judge and autojudging is turned on the judge will send an AJ_AVAILABLE packet.

**Comment by lane55** [comment]
To monitor may need to provide a copy/snapshot of the lists for testing and monitoring.

**Comment by lane55** [comment]
Adding to John's AJ work and plan to push a PR this week.

**Comment by lane55** [comment]
There are additional requirements beyond the basics of assigning runs to a judge.
There are conditions which will cause judges and runs to be added to the AJ lists beyond
run submitted, judge sends AVAILABLE_TO_AUTO_JUDGE packet
Requirement: when run is updated (by admin) to status queued for computer judging add the AJ avail runs
Requirement: on admin when judge is no longer auto-judging remove from AJ avail judges
Requirement: on judge if admin updates auto judging settings, specifically problems to auto judge, must update in AJ avail judge list and perform check to see if judge can be assigned a run to judge
Requirement: admin turns on auto judging for a judge, judge must send a AVAILABLE_TO_AUTO_JUDGE  to server

**Comment by clevengr** [comment]
Just a note to record the fact that the problem being addressed in this issue (#496) is in essence the same problem as what is discussed in Issue #445; see that issue for additional important discussion points.  (The two issues are, in effect, duplicates -- but because of the lengthy discussions in both it's probably not appropriate to close one or the other until the entire issue is resolved...)
